File: python/src/entryconf/_lock.py
"""The directory lock and the atomic replacement write (SPEC §10.7)."""

from __future__ import annotations

import os
import secrets
import sys
import tempfile
import time
from datetime import datetime, timezone
from pathlib import Path

from ._errors import E_LOCKED, E_WRITE, EntryconfError

LOCK_FILE_NAME = ".entryconf.lock"

#: A lock older than this is presumed abandoned by a crashed writer.
LOCK_STALE_AFTER = 30.0

#: How long ``Plan.commit`` waits for the lock by default.
DEFAULT_LOCK_TIMEOUT = 5.0


def acquire_lock(directory: Path, timeout: float) -> Path:
    """Exclusive-create ``<dir>/.entryconf.lock``, retrying until ``timeout``;
    a lock older than :data:`LOCK_STALE_AFTER` is broken by renaming it to a
    unique name and deleting that (so at most one waiter breaks it)."""
    path = directory / LOCK_FILE_NAME
    deadline = time.monotonic() + timeout
    wait = 0.005
    while True:
        try:
            fd = os.open(path, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o644)
        except FileExistsError:
            pass
        except OSError as exc:
            raise EntryconfError(E_LOCKED, f"cannot create lock file {path}: {exc}") from exc
        else:
            with os.fdopen(fd, "w", encoding="utf-8") as f:
                stamp = datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
                f.write(f'{{"pid": {os.getpid()}, "created": "{stamp}"}}\n')
            return path
        try:
            age = time.time() - path.stat().st_mtime
        except OSError:
            continue  # released between our attempts; try again at once
        if age > LOCK_STALE_AFTER:
            stale = Path(f"{path}.{secrets.token_hex(6)}")
            try:
                os.rename(path, stale)
            except OSError:
                pass
            else:
                try:
                    os.unlink(stale)
                except OSError:
                    pass
            continue
        if time.monotonic() > deadline:
            raise EntryconfError(
                E_LOCKED, f"lock file {path} is held by another writer (waited {timeout:g}s)"
            )
        time.sleep(wait)
        wait = min(wait * 2, 0.1)


def release_lock(path: Path) -> None:
    try:
        os.unlink(path)
    except OSError:
        pass


def atomic_write(target: Path, data: bytes) -> None:
    """Replace ``target``'s content: a temporary file in the target's directory,
    flushed, given the target's permission bits, renamed over it. A symlinked
    target is resolved first so the link survives. On any failure the temporary
    file is removed and the target is untouched."""
    resolved = Path(os.path.realpath(target))
    try:
        mode = resolved.stat().st_mode & 0o777
    except OSError as exc:
        raise EntryconfError(E_WRITE, f"cannot stat {resolved}: {exc}") from exc
    try:
        fd, tmp_name = tempfile.mkstemp(dir=str(resolved.parent), prefix=f".{resolved.name}.entryconf-tmp-")
    except OSError as exc:
        raise EntryconfError(E_WRITE, f"cannot create a temporary file beside {resolved}: {exc}") from exc
    tmp = Path(tmp_name)
    try:
        try:
            os.write(fd, data)
            os.fsync(fd)
            if sys.platform != "win32":
                os.fchmod(fd, mode)
        finally:
            os.close(fd)
        os.replace(tmp, resolved)
    except OSError as exc:
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise EntryconfError(E_WRITE, f"cannot replace {resolved}: {exc}") from exc
    # Best effort: make the rename itself durable where the platform allows.
    try:
        dfd = os.open(str(resolved.parent), os.O_RDONLY)
        try:
            os.fsync(dfd)
        finally:
            os.close(dfd)
    except OSError:
        pass

File: rust/src/lock.rs
//! SPEC §10.7 — the directory lock and the atomic replacement write.

use std::fs::{self, File, OpenOptions};
use std::io::Write;
use std::path::{Path, PathBuf};
use std::time::{Duration, Instant, SystemTime};

use crate::error::{Error, ErrorCode};

pub(crate) const LOCK_FILE_NAME: &str = ".entryconf.lock";

/// How old a lock file must be before a waiter may presume its writer crashed.
pub(crate) const LOCK_STALE_AFTER: Duration = Duration::from_secs(30);

/// A held lock; dropping it deletes the lock file.
pub(crate) struct LockGuard {
    path: PathBuf,
}

impl Drop for LockGuard {
    fn drop(&mut self) {
        let _ = fs::remove_file(&self.path);
    }
}

/// Exclusive-creates `<dir>/.entryconf.lock`, retrying with backoff until
/// `timeout` (`E_LOCKED`), breaking a lock older than [`LOCK_STALE_AFTER`] by
/// renaming it to a unique name first so at most one waiter breaks it.
pub(crate) fn acquire(dir: &Path, timeout: Duration) -> Result<LockGuard, Error> {
    let path = dir.join(LOCK_FILE_NAME);
    let deadline = Instant::now() + timeout;
    let mut wait = Duration::from_millis(5);
    loop {
        match OpenOptions::new().write(true).create_new(true).open(&path) {
            Ok(mut file) => {
                let _ = writeln!(
                    file,
                    "{{\"pid\": {}, \"created\": \"{}\"}}",
                    std::process::id(),
                    unix_time_rfc3339()
                );
                return Ok(LockGuard { path });
            }
            Err(e) if e.kind() != std::io::ErrorKind::AlreadyExists => {
                return Err(Error::new(
                    ErrorCode::Locked,
                    format!("cannot create lock file {}: {e}", path.display()),
                ));
            }
            Err(_) => {}
        }
        let is_stale = fs::metadata(&path)
            .and_then(|m| m.modified())
            .ok()
            .and_then(|m| SystemTime::now().duration_since(m).ok())
            .is_some_and(|age| age > LOCK_STALE_AFTER);
        if is_stale {
            let stale = path.with_extension(format!("lock.{}", random_suffix()));
            if fs::rename(&path, &stale).is_ok() {
                let _ = fs::remove_file(&stale);
            }
            continue;
        }
        if Instant::now() >= deadline {
            return Err(Error::new(
                ErrorCode::Locked,
                format!(
                    "lock file {} is held by another writer (waited {timeout:?})",
                    path.display()
                ),
            ));
        }
        std::thread::sleep(wait);
        if wait < Duration::from_millis(100) {
            wait *= 2;
        }
    }
}

fn unix_time_rfc3339() -> String {
    // Informational only; a civil-time rendering without a chrono dependency.
    let secs = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .map_or(0, |d| d.as_secs());
    let days = secs / 86_400;
    let (h, m, s) = ((secs % 86_400) / 3600, (secs % 3600) / 60, secs % 60);
    // Civil-from-days (Howard Hinnant's algorithm).
    let z = days as i64 + 719_468;
    let era = z.div_euclid(146_097);
    let doe = z.rem_euclid(146_097);
    let yoe = (doe - doe / 1460 + doe / 36_524 - doe / 146_096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let mo = if mp < 10 { mp + 3 } else { mp - 9 };
    let y = if mo <= 2 { y + 1 } else { y };
    format!("{y:04}-{mo:02}-{d:02}T{h:02}:{m:02}:{s:02}Z")
}

pub(crate) fn random_suffix() -> String {
    // Enough entropy to make sibling temp names distinct; no crate needed.
    let nanos = SystemTime::now()
        .duration_since(SystemTime::UNIX_EPOCH)
        .map_or(0, |d| d.subsec_nanos());
    let addr = &nanos as *const u32 as usize;
    format!("{:x}{:x}{:x}", std::process::id(), nanos, addr & 0xffff)
}

/// Replaces `target`'s content with `data` (SPEC §10.7 step 3): a temporary
/// file beside the resolved target, flushed, given the target's permission
/// bits, and renamed over it. A symlinked target is resolved first so the link
/// survives. On any failure the temporary file is removed and the target is
/// untouched.
pub(crate) fn atomic_write(target: &Path, data: &[u8]) -> Result<(), Error> {
    let write_err = |step: &str, e: std::io::Error, path: &Path| {
        Error::new(ErrorCode::Write, format!("{step} {}: {e}", path.display()))
    };
    let resolved = fs::canonicalize(target).map_err(|e| write_err("cannot resolve", e, target))?;
    let meta = fs::metadata(&resolved).map_err(|e| write_err("cannot stat", e, &resolved))?;
    let dir = resolved
        .parent()
        .map_or_else(|| PathBuf::from("."), Path::to_path_buf);
    let base = resolved
        .file_name()
        .map(|n| n.to_string_lossy().into_owned())
        .unwrap_or_default();
    let tmp_path = dir.join(format!(".{base}.entryconf-tmp-{}", random_suffix()));

    let mut tmp = OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(&tmp_path)
        .map_err(|e| write_err("cannot create a temporary file beside", e, &resolved))?;

    let result = (|| -> std::io::Result<()> {
        tmp.write_all(data)?;
        tmp.sync_all()?;
        #[cfg(unix)]
        tmp.set_permissions(meta.permissions())?;
        #[cfg(not(unix))]
        let _ = &meta;
        Ok(())
    })();
    drop(tmp);
    if let Err(e) = result {
        let _ = fs::remove_file(&tmp_path);
        return Err(write_err("cannot write replacement for", e, &resolved));
    }
    if let Err(e) = fs::rename(&tmp_path, &resolved) {
        let _ = fs::remove_file(&tmp_path);
        return Err(write_err("cannot replace", e, &resolved));
    }
    // Best effort: make the rename itself durable where the platform allows.
    #[cfg(unix)]
    if let Ok(d) = File::open(&dir) {
        let _ = d.sync_all();
    }
    #[cfg(not(unix))]
    let _ = File::open(&dir);
    Ok(())
}
